File: packages/exporter/src/domain/section-status.ts
import type { SnapshotSection, SnapshotSectionStatus, SnapshotWarning } from "@aa/types";

export function createSection<T>(
  status: SnapshotSectionStatus,
  data?: T,
  warnings?: SnapshotWarning[],
  todos?: string[],
): SnapshotSection<T> {
  const section: SnapshotSection<T> = { status };
  if (typeof data !== "undefined") section.data = data;
  if (warnings && warnings.length > 0) section.warnings = warnings;
  if (todos && todos.length > 0) section.todos = todos;
  return section;
}

export function markNotImplemented<T>(todo: string): SnapshotSection<T> {
  return createSection<T>("not_implemented", undefined, undefined, [todo]);
}

export function markMissingData<T>(warning: SnapshotWarning): SnapshotSection<T> {
  return createSection<T>("missing_data", undefined, [warning]);
}

export function isSectionAvailable<T>(section: SnapshotSection<T>): section is SnapshotSection<T> & { data: T } {
  return section.status === "available" && typeof section.data !== "undefined";
}

File: apps/api/app/modules/export/section_status.py
"""
Helpers for snapshot section statuses.
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional


def create_section(
    status: str,
    data: Any = None,
    warnings: Optional[List[Dict[str, Any]]] = None,
    todos: Optional[List[str]] = None,
) -> Dict[str, Any]:
    section: Dict[str, Any] = {"status": status}
    if data is not None:
        section["data"] = data
    if warnings:
        section["warnings"] = warnings
    if todos:
        section["todos"] = todos
    return section


def mark_not_implemented(todo: str) -> Dict[str, Any]:
    return create_section("not_implemented", todos=[todo])


def mark_missing_data(warning: Dict[str, Any]) -> Dict[str, Any]:
    return create_section("missing_data", warnings=[warning])


def is_section_available(section: Optional[Dict[str, Any]]) -> bool:
    return bool(section) and section.get("status") == "available" and "data" in section

